File: leetcode/49-group-anagrams.py
# https://leetcode.com/problems/group-anagrams/ (top-interview-questions)

from typing import List
from collections import defaultdict

def groupAnagrams(strs: List[str]) -> List[List[str]]:
	buckets = defaultdict(list)
	for string in strs:
		# we give every anagrame of the same group the same hash
		hash = ''.join(sorted(string))
		buckets[hash].append(string)

	return buckets.values()

File: leetcode/49-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
  const anagramsGroups = {};

  for (const str of strs) {
    const sortedStr = str.split('').sort().join('');
    if (!anagramsGroups.hasOwnProperty(sortedStr)) {
      anagramsGroups[sortedStr] = [];
    }
    anagramsGroups[sortedStr].push(str);
  }

  return Object.values(anagramsGroups);
};